README.md
# Python Vs. Zombies 2 (PyVZ2)
Python 3 scripts for modifying Plants Vs. Zombies 2.

## Files:
- LICENCE, this tool is licenced under GPL-3.0, read it for more information.
- OBBEdit, a tool to unpack and patch PVZ2's resources
- SaveMaxer: a tool to generate PVZ2 save files
- README.md, this file

## Credits to these sources for code:
1. Luigi Auriemma for unpacking RSB/1BSR: https://aluigi.altervista.org/bms/1bsr_pgsr.bms
2. Small Pea for the AES encryption: https://smallpeashared.lanzoul.com/iLqI3ze9qrg
3. 1Zulu for the RTON conversion: https://drive.google.com/file/d/0B-SMAZsDM9ERTGwxVkpoN0U0LVE

## Credits to these sources for documentation:
1. Watto Studios for documenting RSB/1BSR: http://www.watto.org/specs.html?specs=Archive_RSB_1BSR
2. YingFengTingYu for documenting the RSB.SMF format: https://github.com/YingFengTingYu/PopStudio
3. TwinKleS-C for documenting RSB/1BSR: https://github.com/TwinKleS-C/PvZTool/blob/main/src/pvztool.c
4. H3x4n1um for the documentating the RTON format: https://github.com/h3x4n1um/RETON

## MODDING SUPPORT
https://discord.gg/CVZdcGKVSw

## REPORT ISSUES ON GITHUB:
https://github.com/Nineteendo/PVZ2tools


changelog.md
# Python VS. Zombies 2 (PyVZ2)
Python3 scripts for modifying Plants VS Zombies 2

# Release 1.0.0 (Jun 24, 2021)
## Master 1.0.0 (Jun 24, 2021)
### Beta 0.1.0 (Jul 11, 2020)
1. DIFF TABLE:
	* ~~Make changelog from a specific depth of a json file~~ (Moved to https://github.com/Nineteendo/DataTools)
		* [Beta 1.0.2b](#beta-1.0.2b-jan-8-2022)
		* [Beta 1.0.2b](#beta-1.0.2b-\(jan-8-2022\))
		* [Beta 1.0.2b](#beta-1.0.2b-\(jan-8,-2022\))
2. JSON WALKER:
	* ~~Get depth from JSON file for DIFF TABLE~~ (Moved to https://github.com/Nineteendo/DataTools for Beta 1.0.2b)
3. RTONS PARSER:
	* **Forked from RTONParser v0.01 by 1Zulu** http://drive.google.com/file/d/0B-SMAZsDM9ERTGwxVkpoN0U0LVE (Jan 8, 2018)
	* Convert .rton to .json
	* No command line arguments
### Beta 0.2.0 (May 18, 2021)
1. JSONS PARSER:
	* Convert .json to .rton
2. RTONS PARSER:
	* Latin strings & duplicate keys
### Beta 0.2.1 (May 24, 2021)
1. JSONS PARSER:
	* Encode keys correctly (Keys could be encoded as RTID instead of strings)
### Beta 0.2.2 (May 26, 2021)
1. JSONS PARSER:
	* More integer types & Unicode
2. RTONS PARSER:
	* Supported unicode strings
### Beta 0.2.3 (May 27, 2021)
1. JSONS PARSER:
	* Disabled other integer types (The game doesn't seem to use them in the packages)
### Beta 0.2.4 (Jun 2, 2021)
1. OBBPatcher:
	* **Forked from OBBPatcher v0.05b by 1Zulu** http://drive.google.com/file/d/0B_u59SyYZBrTMExscEpLdlNFVWc (Oct 8, 2021)
	* Extract & Patch sections from OBBs
	* Configured untill PVZ2 v7.8.1
2. README:
	* Information about the tools
3. RTONS PARSER:
	* ~~Development tool for No_backup~~ (Removed in Beta 1.0.2)
	* Supported RTID type 0 and 2, all negative varints
### Beta 0.2.5 (Jun 3, 2021)
1. JSONS PARSER:
	* More RTID types
	* Automatic .dat & .bin extensions
2. README:
	* Markdown format
### Beta 0.2.6 (Jun 4, 2021)
1. JSONS PARSER:
	* ~~No Backup SHOULD be supported~~ (Reverted in Beta 0.2.7)
	* ~~Output file with problematic keys that are configured~~ (Reverted in Beta 0.2.7)
### Beta 0.2.7 (Jun 5, 2021)
1. JSONS PARSER:
	* Reverted changes from Beta 0.2.6
### Beta 0.2.8 (Jun 7, 2021)
1. JSONS PARSER:
	* Temporary solution for No_Backup files
2. RTONS PARSER:
	* More SilentErrors
	* ~~Re-RTON checker~~ (deleted in Beta 1.0.2)
### Beta 1.0.0 (Jun 24, 2021)
1. JSONS PARSER:
	* Removed jsons.py
	* Improved consistency
	* Input file & Directory
	* Open fail.txt
	* Version Mention & Author
2. OBBPatcher:
	* Python version
	* Configured untill PVZ2 v8.8.1
3. RTONS PARSER:
	* Removed rtons.py
	* Input file & Directory
	* Open fail.txt
	* Version Mention & Author
# Release 1.1.0 (Feb 6, 2022)
## Master 1.1.0 (Feb 6, 2022)
### Beta 1.0.0b (Nov 26, 2021)
1. RTONS PARSER:
	* **Merged with JSONS PARSER**
### Beta 1.0.1 (Jan 3, 2022)
1. OBBUnpacker:
	* **Forked from 1bsr pgsr v0.2.2 by Luigi Auriemma** http://aluigi.altervista.org/bms/1bsr_pgsr.bms (Jun 29, 2015)
	* Unpack obb
	* Fixed path names
### Beta 1.0.1b (Jan 4, 2022)
1. OBBUnpacker:
	* Unpack multiple files
	* Fixed short paths
	* Options
### Beta 1.0.2 (Jan 8, 2022)
1. OBBPatcher:
	* Removed java version
	* README
2. OBBUnpacker:
	* README
3. README:
	* Added OBBUnpacker
4. RTON Converter:
	* Renamed from RTONS PARSER
	* Removed Development tool for No_backup (added in Beta 0.2.4)
	* Removed re-RTON check (added in )
	* No more manually chosen data types
	* Options
	* README
### Beta 1.0.2b (Jan 8, 2022)
1. DIFF TABLE:
	* Moved to DataTools
2. JSON WALKER:
	* Moved to DataTools
2. LICENSE:
	* This tools are licenced under the GPL-3.0 LICENSE
### Beta 1.0.3 (Jan 20, 2022)
1. OBBPatcher:
	* Extracting & Patching based on OBBUnpacker
	* Better path input
	* Options
	* README
2. OBBUnpacker:
	* Improved extraction
	* Better path input
	* More Extensions
	* Options:
		* Added confirmPath
		* More extensions
		* Added pgsrEndswith & pgsrEndswithIgnore
		* Added pgsrStartswith & pgsrStartswithIgnore
	* README:
		* Added confirmPath
		* Added pgsrEndswith & pgsrEndswithIgnore
		* Added pgsrStartswith & pgsrStartswithIgnore
		* Added pgsr_extract
3. README:
	* Removed DIFF TABLE & JSON WALKER
	* **Merged JSONS_PARSER & RTONS_PARSER**
4. RTON Converter:
	* Re-enabled other integer types
	* Better path input
	* Options:
		* Added cachLimit
		* Removed UncachedStrings
	* README:
		* Removed UncachedStrings
### Beta 1.0.3b (Jan 21, 2022)
1. OBBPatcher:
	* Load options with executable
2. OBBUnpacker:
	* Removed 1bsr_pgsr.bms
	* Windows path names
	* Load options with executable
3. RTON Converter:
	* Load options with executable
### Beta 1.0.4 (Jan 22, 2022)
1. OBBEdit:
	* Renamed from OBBUnpacker
	* Fixed wrong size
2. RTON Converter:
	* Increased efficiency
	* Options:
		* Split up cachLimit in cachKeyLimit & cachValueLimit
	* README:
		* Split up cachLimit in cachKeyLimit & cachValueLimit
### Beta 1.0.5 (Jan 29, 2022)
1. OBBEdit:
	* **Merged with OBBPatcher**
	* Added patching files & sections
	* Options:
		* Split up extractRsgp in rsbUnpackLevel & rsgpUnpackLevel
		* Split up extensions in rsbExtensions & rsgpExtensions
		* Renamed some options
	* README:
		* Split up extractRsgp in rsbUnpackLevel & rsgpUnpackLevel
		* Split up extensions in rsbExtensions & rsgpExtensions
		* Renamed some options
		* rsgp_extract
2. README:
	* **Merged OBBPatcher & OBBUnpacker**
### Beta 1.0.5b (Jan 30, 2022)
1. OBBEdit:
	* Improved error log
2. RTON Converter:
	* Improved error log
### Beta 1.0.5c (Jan 31, 2022)
1. RTON Converter:
	* Improved objects
### Beta 1.0.5d (Feb 1, 2022)
1. RTON Converter:
	* Enforce Python 3.9 after a report of mutaqin-hanif on GitHub http://github.com/Nineteendo/PVZ2tools/issues/4 (Feb 1, 2022)
	* Fixed short names
### Beta 1.0.6 (Feb 2, 2022)
1. OBBEdit:
	* Options:
		* Enabled DEBUG MODE
		* Removed dumpRsgp
	* README:
		* Removed dumpRsgp
2. RTON Converter:
	* Faster conversions
	* Options:
		* Added sortValues
		* Removed allowNan & ensureAscii
		* Renamed some options
	* README:
		* Added sortValues
		* Removed allowNan & ensureAscii

SaveMaxer/max.py
from datetime import datetime
from os.path import join as osjoin
from json import dump
from struct import pack

# 3th party libraries
from libraries.pyvz2nineteendo import ProgressBar, LogError
from libraries.pyvz2rton import JSONDecoder

options = {
# Default options
	"prefix": "AHZ",
	"profiles": []
}

def member(member, check):
	return type(member) == check
def get_list(start = 1, last_id = 0):
	if member(last_id, list):
		return last_id
	else:
		return list(range(start, last_id + 1))
def get_uid(generated_uid = True, prefix = "", id = 0):
	if member(generated_uid, str):
		return generated_uid
	else:
		a, b = divmod(hash(get_n(generated_uid, prefix, id)), 4294967296)
		b, c = divmod(b, 65536)
		return repr(b) + "." + repr(c) + "." + pack(">i", a).hex()
def get_v(enable_gems = True):
	if member(enable_gems, int):
		return enable_gems
	elif enable_gems:
		return 9 # Allow gems to be set in the stats
	else:
		return 8 # Prevent gems to be set in the stats
def get_n(generated_name = True, prefix = "", id = 0):
	if member(generated_name, str):
		return generated_name
	elif generated_name:
		return prefix + repr(id)
	else:
		return prefix
def get_i(accept_updated_terms = True, i = 0):
	if member(accept_updated_terms, int):
		return accept_updated_terms
	elif accept_updated_terms:
		return int(datetime.timestamp(datetime.now())) + i # # Disable updated terms message
	else:
		return 0 # Enable updated terms message
def get_p(plants = 256):
	if member(plants, list):
		return plants
	else:
		removed_plants = [43, 45, 46, 47, 48, 49, 56, 57, 73]
		return [plant for plant in get_list(1, plants) if not plant in removed_plants]
def get_wm(old_world_map = {}):
	if member(old_world_map, list):
		return old_world_map
	else:
		wm = []
		for world_id in old_world_map:
			for level_id in get_list(1, old_world_map[world_id]):
				level = {}
				level["W"] = int(world_id) # World ID
				level["E"] = level_id # Level ID
				level["S"] = 3 # Unknown
				level["C"] = 7 # Unknown
				level["N"] = 0 # Unknown
				level["G"] = 0 # Unknown
				
				wm.append(level)
		return wm
def get_wmed(new_world_map = {}):
	if member(new_world_map, list):
		return new_world_map
	else:
		wmed = []
		for world_id in new_world_map:
			world = {}
			world["w"] = int(world_id) # World ID
			
			e = []
			for level_id in get_list(1, new_world_map[world_id]):
				level = {}
				level["i"] = level_id # Level ID
				e.append(level)
			
			world["e"] = e # Levels
			world["r"] = False # Unknown
			
			wmed.append(world)
		return wmed
def get_dri(danger_rooms = {}):
	if member(danger_rooms, list):
		return danger_rooms
	else:
		dri = []
		for world_name in danger_rooms:
			world_options = danger_rooms[world_name]
			dangerroom = {}
			dangerroom["wn"] = world_name # World Name
			dangerroom["cl"] = world_options["current_level"] # Current Level
			dangerroom["hl"] = world_options["high_score"] # High Score
			dangerroom["pfco"] = world_options["plant_food"] # Plantfood
			dangerroom["l"] = world_options["lawn_mowers"] # Lawn Mowers https://images.wikia.com/ernestoam/images/b/b1/Endless_Zone_Lawn_Mower_Set_Up_IDs.png
			dangerroom["sb"] = get_p(world_options["plants"]) # Plants
			dangerroom["ts"] = 0 # Unknown
			dangerroom["pl"] = 0 # Unknown
			dangerroom["hr"] = world_options["draw_card"] # Draw a Card
			dangerroom["hm"] = False # Unknown
			
			dri.append(dangerroom)
		return dri
def get_pr(power_up_counts = {}):
	if member(power_up_counts, list):
		return power_up_counts
	else:
		pr = []
		for powerup_name in power_up_counts:
			powerup = {}
			powerup["n"] = powerup_name # Powerup name
			powerup["i"] = power_up_counts[powerup_name]
			
			pr.append(powerup)
		return pr
def get_tyi(level_name = "egypt1"):
	if member(level_name, dict):
		return level_name
	else:
		tyi = {}
		tyi["wml"] = level_name # Level Name
		tyi["lst"] = 0 # Unknown
		tyi["nst"] = 0 # Unknown
		return tyi
def get_pbi(boosted_plants = 256):
	if member(boosted_plants, list):
		return boosted_plants
	else:
		unboosted_plants = [29, 30, 32, 33, 41, 68, 72, 79, 89, 94, 97, 111, 118, 124, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 145, 155, 181, 184, 188]
		pbi = []
		for boosted_plant in range(1, boosted_plants + 1):
			if not boosted_plant in unboosted_plants:
				plant = {}
				plant["p"] = boosted_plant # Plant ID
				plant["c"] = 127 # Unknown
				plant["a"] = 4 # Boost Type https://pastebin.com/tpZaxBgX
				plant["b"] = [] # Unknown
				plant["t"] = 0 # Next time accessible
				
				pbi.append(plant)
		return pbi
def get_upsi(bought_plant_pots = 6):
	if member(bought_plant_pots, list):
		return bought_plant_pots
	else:
		bought_pot_ids = [201, 301, 2, 102, 202, 302]
		return sorted([bought_pot_ids[extra_pot] for extra_pot in range(0, bought_plant_pots)])
def get_mcs(enable_treasure_yeti = True):
	if member(enable_treasure_yeti, int):
		return enable_treasure_yeti
	elif enable_treasure_yeti:
		return 3
	else:
		return 2 # Unknown
def get_ap(played_mini_games = {}):
	if member(played_mini_games, list):
		return played_mini_games
	else:
		ap = []
		for played_pack in played_mini_games["mini_games"]:
			pack = {}
			pack["i"] = int(played_pack) # Mini Game ID
			
			lp = []
			for level_id in played_mini_games["mini_games"][played_pack]:
				level = {}
				level["i"] = level_id # Level ID
				level["p"] = played_mini_games["wave"] # Wave
				level["s"] = 1 # Unknown
				
				lp.append(level)
			
			pack["lp"] = lp
			
			ap.append(pack)
		return ap
def get_up(collected_rewards = {}):
	if member(collected_rewards, list):
		return collected_rewards
	else:
		ap = []
		for minigame_id in collected_rewards:
			minigame = {}
			minigame["i"] = int(minigame_id) # Mini Game ID
			
			up = []
			for played_pack in collected_rewards[minigame_id]:
				pack = {}
				pack["i"] = played_pack # pack ID
				pack["p"] = 0 # Unknown
				
				up.append(pack)
			
			minigame["up"] = up
			
			ap.append(minigame)
		return ap
def get_plis(upgraded_plants = 256):
	if member(upgraded_plants, list):
		return upgraded_plants
	else:
		plis = []
		for upgraded_plant in get_p(upgraded_plants["plants"]):
			plant = {}
			plant["p"] = upgraded_plant # Plant ID
			plant["l"] = upgraded_plants["level"] # Plant Level
			plant["x"] = upgraded_plants["seed_packets"] # Seed mini_gameets
			plant["m"] = upgraded_plants["mastery"] # Plant Mastery
			
			plis.append(plant)
		return plis
def generate_save_file():
	json_data = {}
	json_data["version"] = 1

	objects = []
	entries = options["profiles"]
	split_task(len(entries))
	for i, profile_options in enumerate(entries):
		profile = {}
		profile["uid"] = get_uid(profile_options["generated_uid"], options["prefix"], i) # Uid
		profile["objclass"] = "PlayerInfo" # Player Info
		
		objdata = {}
		objdata["v"] = get_v(profile_options["enable_gems"])
		objdata["n"] = get_n(profile_options["generated_name"], options["prefix"], i) # Name
		objdata["i"] = get_i(profile_options["accept_updated_terms"], i)
		objdata["l"] = profile_options["last_level"] # Last level
		objdata["c"] = profile_options["coins"] # Coins
		objdata["g"] = profile_options["gems"] # Gems
		objdata["t"] = profile_options["gauntlets"] # Gauntlets
		objdata["m"] = profile_options["mints"] # Mints
		objdata["pf"] = profile_options["penny_fuel"] # Penny Fuel
		#objdata["pt"] = 888888888 # Perk Progression
		objdata["p"] = get_p(profile_options["plants"]) # Plants https://pastebin.com/bSeTQTRa https://pastebin.com/JyCbwEq2
		objdata["kz"] = get_list(0, profile_options["killed_zombies"]) # Killed Zombies https://pastebin.com/1WhUWS1Q
		objdata["wm"] = get_wm(profile_options["old_world_map"]) # Old World Map
		objdata["wmed"] = get_wmed(profile_options["new_world_map"]) # New World Map
		objdata["gf"] = get_list(0, profile_options["game_features"]) # Game Features https://pastebin.com/W1hSXBPB
		objdata["ne"] = get_list(1, profile_options["narrations"]) # Narrations https://pastebin.com/hHvqxxr2
		objdata["ct"] = get_list(1, profile_options["travel_log_narrations"]) # Travel Log
		objdata["dri"] = get_dri(profile_options["danger_rooms"]) # Dangerroom
		objdata["cos"] = get_list(0, profile_options["costumes"]) # Costumes https://pastebin.com/RG41dYRv
		objdata["pr"] = get_pr(profile_options["power_up_counts"]) # Powerups
		objdata["tyi"] = get_tyi(profile_options["treasure_yeti"]) # Treasure Yeti In
		objdata["pbi"] = get_pbi(profile_options["boosted_plants_backup"]) # Boosted Plant Backup?
		objdata["gpi"] = profile_options["zen_garden_plants"] # Zen Garden Plants
		objdata["upsi"] = get_upsi(profile_options["bought_plant_pots"]) # Plant Pots
		objdata["spr"] = profile_options["sprouts"] # Sprouts
		objdata["pli"] = get_pbi(profile_options["boosted_plants"]) # Boosted Plants
		objdata["izg"] = profile_options["start_from_zen_garden"] # Start from Zen Garden
		objdata["mcs"] = get_mcs(profile_options["enable_treasure_yeti"]) # Enable Treasure Yeti
		objdata["ap"] = get_ap(profile_options["played_mini_games"]) # Played Mini Games
		objdata["puc"] = get_up(profile_options["collected_rewards"]) # Collected rewards
		objdata["plis"] = get_plis(profile_options["upgraded_plants"]) # Plant Leveling

		profile["objdata"] = objdata # Profile Data
		
		objects.append(profile)
		finish_sub_task()
	json_data["objects"] = objects # Profiles
	merge_task()
	return json_data

# Start of the code
try:
	logerror = LogError()
	error_message = logerror.error_message
	input_level = logerror.input_level
	warning_message = logerror.warning_message
	application_path = logerror.application_path

	logerror.check_version(3, 9, 0)
	branches = {
		"beta": "Beta 1.2.2c fixed patching, json encode & progressbar",
		"master": "Merge branch 'beta'"
	}
	release_tag = "1.2"
	print("""\033[95m
\033[1mSaveMaxer v1.2.2c (c) 2022 Nineteendo\033[22m
\033[1mFollow PyVZ2 development:\033[22m \033[4mhttps://discord.gg/CVZdcGKVSw\033[24m
\033[0m""")
	getupdate = logerror.get_update("Nineteendo/PVZ2tools", branches, release_tag)
	options = logerror.load_template(options, 2)
	encode_root_object = JSONDecoder().encode_root_object

	progressbar = ProgressBar(levels = 2, fail = logerror.fail)
	
	split_task = progressbar.split_task
	merge_task = progressbar.merge_task
	finish_sub_task = progressbar.finish_sub_task
	split_task(3)
	json_data = generate_save_file()
	finish_sub_task()
	dump(json_data, open(osjoin(application_path, "pp.dat.json"), "w"), ensure_ascii = False, indent = 4)
	finish_sub_task()
	patch_data = JSONDecoder().encode_root_object(open(osjoin(application_path, "pp.dat.json"), "rb"))
	open(osjoin(application_path, "pp.dat"), "wb").write(patch_data)
	finish_sub_task()
	
	logerror.finish_program()
except Exception as e:
	error_message(e)
except BaseException as e:
	warning_message(type(e).__name__ + " : " + str(e))
logerror.close() # Close log